scalafix-core/shared/src/main/scala/scalafix/internal/rule/MissingFinal.scala

```diff
@@ -38,10 +39,24 @@ final case class MissingFinal(index: SemanticdbIndex)
           }
       }
 
+    def isNonFinalConcreteCaseClass(mods: List[Mod]) =
+      mods.exists(_.is[Mod.Case]) &&
+        !mods.exists(m => m.is[Mod.Final] || m.is[Mod.Abstract])
+
+    def holdsOuterReference(t: Defn.Class) = {
+      @tailrec
+      def loop(parent: Option[Tree]): Boolean = parent match {
+        case Some(t: Template) => loop(t.parent)
+        case Some(t: Defn.Object) => loop(t.parent)
+        case Some(_: Defn.Class) | Some(_: Defn.Trait) => true
+        case _ => false
+      }
+      loop(t.parent)
+    }
+
     ctx.tree.collect {
       case t @ Defn.Class(mods, _, _, _, _)
-          if mods.exists(_.is[Mod.Case]) &&
-            !mods.exists(_.is[Mod.Final]) =>
+          if isNonFinalConcreteCaseClass(mods) && !holdsOuterReference(t) =>
```

Comment:
That's not very good to traverse tree upwards, `collect` goes from root to leaves. You can look at the `ContextTraverser` in the `Disable` rule for ideas how not to use `.parent` method, but I suspect it will be a more knotty solution than this. So I don't require to rewrite it, it's up to you.

Reply:
that's a good point 👍. I reimplemented the rewrite following your suggestion but the result turned out to be a lot clumsy. I will see if I it's possible to cut down some of the boilerplate. If not, I wonder whether the improved efficiency outweighs the additional complexity. The tree should be very shallow where case classes are usually defined anyway.

Reply:
@vovapolu, I just updated the PR. I drew inspiration from `ContextTraverser` but found it simpler/concise (~20 lines shorter) to update the indicator of potential outer references from within the traversal rather than in the PF (avoids some back and forth). Please let me know if you are ok with that, have other suggestions or prefer to rollback to the previous solution.

Reply:
I withdraw my words... I found a cleaner way